systemctl: add service-log-{level,target} verbs

keszybz opened the pull request:
Heavily inspired by #15622. This adds:
  systemctl service-log-level systemd-resolved
  systemctl service-log-level systemd-resolved info
  systemctl service-log-target systemd-resolved
  systemctl service-log-target systemd-resolved console

We already have systemctl verbs log-level, log-target, and service-watchdogs.
Those two new verbs tie nicely into this scheme.

yuwata commented:
Sorry, I have not followed the previous discussion. But I have a question. Is this meaningful only for systemd services, like systemd-resolved, journald, networkd, etal?

keszybz commented:
> Sorry, I have not followed the previous discussion. But I have a question. Is this meaningful only for systemd services, like systemd-resolved, journald, networkd, etal?

It is applicable to any service that implements the [interface](https://www.freedesktop.org/software/systemd/man/org.freedesktop.LogControl1.html). Right now that's only our services, but there might be more in the future.

yuwata commented on man/systemctl.xml:
```diff
@@ -1120,6 +1120,58 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
+        <varlistentry>
+          <term><command>service-log-level</command> <replaceable>SERVICE</replaceable> [<replaceable>LEVEL</replaceable>]</term>
+
+          <listitem><para>If no argument is given, print the current log level of the given service
```
`no argument` sounds sightly strange for me, as it requires `SERVICE` argument.

yuwata commented on man/systemctl.xml:
```diff
@@ -1120,6 +1120,58 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
+          <replaceable>name</replaceable>.)</para>
+
+          <para>If an optional argument <replaceable>LEVEL</replaceable> is provided, then the command
+          changes the current log level of the service to <replaceable>LEVEL</replaceable>.  The log level
```
Not a matter, but white space is doubled.

yuwata commented on man/systemctl.xml:
```diff
@@ -1120,6 +1120,58 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
+        <varlistentry>
+          <term><command>service-log-target</command> <replaceable>SERVICE</replaceable> [<replaceable>TARGET</replaceable>]</term>
+
+          <listitem><para>If no argument is given, print the current log target of the given service
```
Same here.

yuwata commented on src/systemctl/systemctl.c:
```diff
@@ -6326,6 +6326,119 @@ static int log_level(int argc, char *argv[], void *userdata) {
+                                &error,
+                                &names);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine Names property of %s: %s",
```
`Names=`?

Maybe, `determine` -> `obtain` or `get`?

yuwata commented on src/systemctl/systemctl.c:
```diff
@@ -6326,6 +6326,119 @@ static int log_level(int argc, char *argv[], void *userdata) {
+
+        q = endswith(p, ".service");
+        if (!q)
+                return log_error("%s is not a service, cannot set log level for it.", p);
```
This should return negative errno. Also, `cannot set log level for it.` should be dropped or updated, as this is also used for `systemctl service-log-target`.

keszybz commented:
Thank you for the review. Updated with all the suggested changes.

yuwata commented on man/systemctl.xml:
```diff
@@ -1120,6 +1120,58 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
+        <varlistentry>
+          <term><command>service-log-target</command> <replaceable>SERVICE</replaceable> [<replaceable>TARGET</replaceable>]</term>
+
+          <listitem><para>If the <replaceable>TARGET</replaceable> argument is given, print the current log
```
Missing 'no'.

keszybz replied:
Oops, fixed. (I also made the descriptions for service-log-level and service-log-target use the same sentence structure. The error was partially caused by them being different.)

yuwata reviewed (changes requested):
Looks mostly good for me, and I prefer this instead of implementing the same in `busctl`. But one comment. See below.

yuwata commented:
LGTM. But let's wait for comments by others, @poettering?

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -7773,6 +7887,8 @@ static int systemctl_help(void) {
                "  log-level [LEVEL]                   Get/set logging threshold for manager\n"
                "  log-target [TARGET]                 Get/set logging target for manager\n"
+               "  service-log-level SERVICE [LEVEL]   Get/set logging threshold for service\n"
+               "  service-log-target SERVICE [TARGET] Get/set logging target for service\n"
```
grmbl. yet another debugging verb in systemctl. I am not a fan of that. I wish we could at least fold this into the existing  "log-level" verb, like we do for "show" where you can specify a unit but if you don#t the call applies to the manager...

But the ambiuity kills this... i.e. having a service called "error.service" would mean "systemctl log-level error" would not have a clear meaning anymore...

poettering replied:
Why is this under "Manager State Commands" section? it's got nothing to do with the manager state, it's a per service thing

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -7773,6 +7887,8 @@ static int systemctl_help(void) {
                "  log-target [TARGET]                 Get/set logging target for manager\n"
+               "  service-log-level SERVICE [LEVEL]   Get/set logging threshold for service\n"
+               "  service-log-target SERVICE [TARGET] Get/set logging target for service\n"
                "  service-watchdogs [BOOL]            Get/set service watchdog state\n"
```
hmm, it kinda sucks that this is called "service-watchdogs" even though it applies to all services... i.e. some "service" verbs now apply to specific services, while this one applies to all them.

This really feels like we really need to clean up some naming here...

poettering replied:
maybe the "log-level" verbs should actually be called "daemon-log-level" or so, since we use "daemon" as a moniker for referencing the manager itself.

("daemon" is a shitty moniker for that, should probably have been "manager" i guess, but "systemctl daemon-reload" one of the primary ifaces we advertise, would be superhard to rename that)

keszybz replied:
manager-reload sounds nice. I don't think it's a big issue to change the name: we'll keep backwards compat indefinitely.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6326,6 +6326,120 @@ static int log_level(int argc, char *argv[], void *userdata) {
+                        p = startswith(*s, "dbus-");
+                        if (p)
+                                break;
+                }
```
this is hacky, since not all services actually do dbus activation, but the "dbus-xyz" thing is only needed to dbus activation.

This should query the BusName= property I think

keszybz commented:
I screwed up the rebase on the first try, please ignore the comments from github above.

I moved the commands and changed the logic a bit: BusName is checked first, and then any `dbus-<foo>.service` name that passes `service_name_is_valid()`.  `systemctl show -p Names,BusName '*.service'` shows that there are units which have only BusName and units which only have the dbus-foo name and that there are units which have both.

I also added a commit which sorts the unit verbs in `--help` and the man page: introspection first, changes second. The two new verbs are both, but I put them in the second section. I think this makes the whole thing a bit clearer.

Re renaming verbs: I think we could do that. Another one that sticks out is `get-default`/`set-default`. If we rename it, we could combine the two into one get/set verb. What about `default-target`? (Yes, that is two renames in one. I always found the `systemd.unit=` awkward. I would prefer `systemd.target=`, since a target is almost always used and the fact that a non-target unit may be specified is just a detail.)
And `manager-reload` and `manager-restart` seem quite nice.

keszybz commented:
bionic-* is failing on #16994, but I don't think we need to rerun this, since this new functionality is not tested in any way by the CI, and the general compilation + unittests on other Cis are enough.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6327,6 +6328,137 @@ static int log_level(int argc, char *argv[], void *userdata) {
+                                "org.freedesktop.systemd1.Unit",
+                                "Names",
+                                &error,
+                                &names);
```
I'd really drop the alias thing. we load aliases only if they are requested, and if they are never requested then the `Names` property will not list them. hence it is not a reliable way to figure out bus services. It will work sometimes and sometimes not, depending how the service was first activated. it's a game...

I think we should really make this explicit, i.e. only honour BusName= for this, and add good explanations in the man page in the error message suggesting people to add BusName= to their service if they want this to work. i.e. if you invoke this command and you get back "Your service %s is not a bus service or does not declare BusName= setting. Please add it and implement XYZ". 

I think in the long-run everyone wins if services have BusName= set properly, i.e. carry correct metadata, and this might be the way to get there.

I mean, after all this stuff won't work without explicit support in the service anyway, as the services have to explicitly implement the LogControl1 interface. And if they implement that, then also adding BusName= to the service file is kinda implicit... 

I prepped # now, which adds the missing BusName= to all our own services where it was missing.

So I'd propose:

1. Let's make the man page description of "systemctl service-log-level" explicit, and say that the services need to implement LogControl1 + set BusName= for this to work.
2. Update the BusName= description in the man page, to say that this is useful both for Type=dbus *and* `systemctl service-log-level`
3. Update the org.freedesktop.LogControl1 man page to point to the systemctl verb, and that BusName= should be set
4. Update the systemctl command to give very explicit hints if invoked on services that do not carry BusName= or that do not impement LogControl1, maybe even if an urlified terminal hyperlink to the org.freedesktop.LogControl1 man page.

I think the outcome of this would be best: everything is systematic, easy to discover and well documented

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6327,6 +6328,137 @@ static int log_level(int argc, char *argv[], void *userdata) {
+                if (r < 0)
+                        return log_error_errno(r, "Failed to set log %s of service %s to %s: %s",
+                                               level ? "level" : "target",
+                                               dbus_name, argv[2], bus_error_message(&error, r));
```
i wonder if we should do some client-side mangling or validation of at least the log level part, so that people don't willy-nilly come up with their own level definitions or so. because a generic interface that accepts non-generic parameters is not really generic. 

i.e. maybe either parse the syslog level and always pass it in number form (or only in text form). Or at least warn if it doesn't parse as number/text string. Slight preference for the former, since it means we can remove any potential confusion on whether both lowercase/uppercase should be accepted or whether the log levels in text form should be prefixed with `LOG_` or not.

Doing the same kind of mangling for the target is much harder, since the acceptable vocabulary is less clear.

poettering commented:
I still think we should merge the "busctl" hookup for this too, simply because there are many participants on dbus that are not systemd services necessarily. i.e. i think there's value also for dbus clients (i.e. peers that only have a unique name but no well-known name) to have their log level controlled. Also PID 1 itself is not a service (but exposed as scope init.scope), and it might make sense to export it like that.

keszybz commented:
>     1. Let's make the man page description of "systemctl service-log-level" explicit, and say that the services need to implement LogControl1 + set BusName= for this to work.

Done.

>     2. Update the BusName= description in the man page, to say that this is useful both for Type=dbus _and_ `systemctl service-log-level`

Done.

>     3. Update the org.freedesktop.LogControl1 man page to point to the systemctl verb, and that BusName= should be set

Done.

>     4. Update the systemctl command to give very explicit hints if invoked on services that do not carry BusName= or that do not impement LogControl1, maybe even if an urlified terminal hyperlink to the org.freedesktop.LogControl1 man page.

Done.

>  wonder if we should do some client-side mangling or validation of at least the log level part, so that people don't willy-nilly come up with their own level definitions or so.

Verification is now done for the log level. Only the textual names are allowed, see the commit message.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6380,72 +6394,13 @@ static int service_log_setting(int argc, char *argv[], void *userdata) {
-                if (r < 0)
-                        return log_error_errno(r, "Failed to set log target: %s", bus_error_message(&error, r));
-        }
+        const BusLocator bloc = {
```
i find "const" weird for something that isn't really const. it gets us nothing and is just confusing?

maybe put this together on the fly on the invocation log_setting_internal() below

```c
return log_setting_internal(
                bus,
                &(BusLocator) {
                        .destination = …
                },
                argv[0], argv[2]);
```

keszybz replied:
I know this is possible, but this call is already fairly big. I find it nicer to split in two. And I think `const` is fine: it's just saying that the variable will not be modified afterwards.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6307,6 +6308,12 @@ static int log_setting_internal(sd_bus *bus, const BusLocator* bloc, const char
 
         if (value) {
+                if (level) {
+                        if (log_level_from_string(value) < 0)
```
i'd allow numerical levels too, after all they are standardized pretty much. But either is fine to me. However, the man page in your earlier commit of this PR claims that numbers are accepted too, that part needs to be adjusted hence.

keszybz replied:
OK, let's allow numerical too.

keszybz replied:
Actually, the code was allowing numerical too, so I only needed to adjust the man page ;) I tested that it rejects invalid names and big numbers, but I didn't test single integers ;)

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6299,6 +6299,18 @@ static int switch_root(int argc, char *argv[], void *userdata) {
+        (void) terminal_urlify_man("org.freedesktop.LogControl1", "5", &link);
+
+        log_notice("Hint: the service must declare BusName= and implement the appropriate D-Bus interface.\n"
+                   "      See the %s for details.", link ?: "man page");
```
I think the `the` should move into the "man page" string, otherwise you'll say `See the man: org.freedesktop.LogControl1 for details`, which is weird, the `the` appears wrong in that

keszybz replied:
The generated string works like this:
```
      See the org.freedesktop.LogControl1(5) man page for details.
```

But I'll change the fallback string to include the man page name too. It's all theoretical since it'd only trigger on oom, but let's be fully correct.

keszybz commented:
Updated with the suggested changed (except for the man page string, which I think was mostly correct).

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -6317,20 +6329,29 @@ static int log_setting_internal(sd_bus *bus, const BusLocator* bloc, const char
+                        log_error_errno(r, "Failed to set log %s of %s to %s: %s",
+                                        level ? "level" : "target",
+                                        bloc->destination, value, bus_error_message(&error, r));
+                        if (sd_bus_error_has_name(&error, SD_BUS_ERROR_UNKNOWN_METHOD))
```
hmm, the error returned depends on the bus implementation and app.

Might also be one of SD_BUS_ERROR_UNKNOWN_OBJECT, SD_BUS_ERROR_UNKNOWN_INTERFACE, SD_BUS_ERROR_UNKNOWN_PROPERTY

poettering replied:
maybe a common bus_error_is_unsupported_api() helper might make sense, or so

poettering reviewed (commented):
One dirty dirty hack one could add is to check if the specified unit name is "init.scope", and if so, make "service-log-level" an alias for "log-level"...

keszybz commented:
Updated with the suggested change. I didn't implement the helper, but instead revert the check so that one calls to sd_bus_has_names is enough.

yuwata commented:
Hmm, `test_lldp` randomly fails... Of course, not related to this PR.
```
======================================================================
FAIL: test_lldp (__main__.NetworkdLLDPTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 3101, in test_lldp
    self.assertRegex(output, 'veth-peer')
AssertionError: Regex didn't match: 'veth-peer' not found in 'LINK             CHASSIS ID        SYSTEM NAME      CAPS        PORT ID           PORT DESCRIPTION\n\n0 neighbors listed.'
```
